### ISSUE TITLE
Type annotation for `user_rules` parameter of the `Linter` class is wrong

### ISSUE THREAD

**Issue by Samyak2** [sqlfluff/sqlfluff#3973]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I'm trying out sqlfluff to use as a base to implement my custom linting rules. So I implemented a class inheriting from `BaseRule` and tried to use it with the `Linter` class and the `user_rules` parameter. My editor showed an error when I passed in just the class without instantiating it, but passing an object made it fail in run time.

Side note: the [docs page](https://docs.sqlfluff.com/en/stable/developingrules.html) for developing rules does not mention `groups`, `crawl_behaviour` and other attributes or even how to use the custom rule once it is defined. I had to look at the API reference to find `Linter` (so I'm not entirely sure if there's a better way to register a custom rule).

### Expected Behaviour

The `user_rules` parameter should accept `List[Type[BaseRule]]`

### Observed Behaviour

The `user_rules` parameter accepts `List[BaseRule]` which are instances of a `BaseRule`.

### How to reproduce

Here is a minimal reproducible example:
```python
from typing import List

from sqlfluff.core import Linter
from sqlfluff.core.rules import BaseRule, LintResult, RuleContext
from sqlfluff.core.rules.crawlers import RootOnlyCrawler


class Rule_Something_L001(BaseRule):
    """Does nothing."""

    groups = ("all",)

    crawl_behaviour = RootOnlyCrawler()

    def _eval(self, context: RuleContext) -> List[LintResult]:
        return []


linter = Linter(
    user_rules=[Rule_Something_L001], dialect="ansi"  # the relevant part
)
result = linter.lint_string("SELECT 1")
```

pyright in my editor shows this error on line 20:
```
1. Argument of type "list[Type[Rule_Something_L001]]" cannot be assigned to parameter "user_rules" of type "List[BaseRule] | None" in function "__init__"
     "Type[type]" is incompatible with "Type[BaseRule]"
```

mypy shows this error:
```
repro.py:20: error: List item 0 has incompatible type "Type[Rule_Something_L001]"; expected "BaseRule"
```

Now, if I replace that with an instance of the class, there are type errors but there is a run time error:
```
AttributeError: 'Rule_Something_L001' object has no attribute '__name__'. Did you mean: '__ne__'?
```

---

(couldn't find a section for additional context or potential fixes, so including it here)

I think this was not caught in type checking because the only thing `Linter` does is pass it to `RuleSet.register` [here](https://github.com/sqlfluff/sqlfluff/blob/121d6dd415f7436a375b9a3ec33ffd0a6283b530/src/sqlfluff/core/linter/linter.py#L89) which does not have a type annotation for it's parameter [here](https://github.com/sqlfluff/sqlfluff/blob/121d6dd415f7436a375b9a3ec33ffd0a6283b530/src/sqlfluff/core/rules/base.py#L894). So adding this type annotation should prevent such mistakes in the future.

### Dialect

Reproduced with ANSI and Snowflake

### Version

1.3.2

### Configuration

None

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by barrywhart** [comment]
Did you implement the custom rule as a [plugin](https://docs.sqlfluff.com/en/stable/developingplugins.html)? That's the official way to do it. I think `user_rules` is only used by SQLFluff automated tests which sometimes need to register one-off, oddball rules that no one would use for real (e.g. a rule that deliberately introduces an endless cycle of fixes).

You're right that the docs are out of date. There's been a lot of evolution in the rule framework in the past year. Existing rules are generally the best guide to implementing custom rules. If you are willing to help update the docs, though, that'd be 💯🙌.

**Comment by Samyak2** [comment]
Ah I see. I will try making a plugin instead of hacking on the core like this. Thank you.

Meanwhile, I can still send a PR to fix this particular type annotation. Do you want me to?

**Comment by barrywhart** [comment]
Sure, we like type annotations! Note that the GitHub CI build runs mypy on all PRs, so make sure your changes don't cause any mypy failures.

**Comment by Samyak2** [comment]
`mypy` did pass, I checked from the pre-commit hook
